src/python/turicreate/test/test_sframe_s3.py

```diff
@@ -0,0 +1,127 @@
+import shutil
+import pytest
+import boto3
+
+
+@pytest.mark.skipif(
+    os.environ.get("DTURI_ENABLE_SF_S3") is None,
+    reason="slow IO test; enabled when needed",
+)
+class TestSFrameS3(object):
+    @classmethod
+    def setup_class(self):
+        self.my_tempdir = tempfile.mkdtemp()
+        self.s3_client = boto3.client(
+            "s3",
+            endpoint_url=os.environ["TURI_S3_ENDPOINT"],
+            region_name=os.environ["TURI_S3_REGION"],
+        )
+        self.bucket = "tc_qa"
+        self.s3_prefix = "integration/manual/"
+
+    @classmethod
+    def teardown_class(self):
+        shutil.rmtree(self.my_tempdir)
+
+    def _assert_sarray_equal(self, sa1, sa2):
+        l1 = list(sa1)
+        l2 = list(sa2)
+        self.assertEqual(len(l1), len(l2))
+        for i in range(len(l1)):
+            v1 = l1[i]
+            v2 = l2[i]
+            if v1 is None:
+                self.assertEqual(v2, None)
+            else:
+                if type(v1) == dict:
+                    self.assertEqual(len(v1), len(v2))
+                    for key in v1:
+                        self.assertTrue(key in v1)
+                        self.assertEqual(v1[key], v2[key])
+
+                elif hasattr(v1, "__iter__"):
+                    self.assertEqual(len(v1), len(v2))
+                    for j in range(len(v1)):
+                        t1 = v1[j]
+                        t2 = v2[j]
+                        if type(t1) == float:
+                            if math.isnan(t1):
+                                self.assertTrue(math.isnan(t2))
+                            else:
+                                self.assertEqual(t1, t2)
+                        else:
+                            self.assertEqual(t1, t2)
+                else:
+                    self.assertEqual(v1, v2)
+
+    def _test_sframe_equal(self, sf1, sf2):
+        # asserts two frames are equal, ignoring column ordering.
+        self.assertEqual(sf1.num_rows(), sf2.shape[0])
+        self.assertEqual(sf1.num_columns(), sf2.shape[1])
+        for col in sf1.column_names():
+            self._assert_sarray_equal(sf1[col], sf2[col])
+
+    def test_s3_csv(self):
+        fname = os.path.join(self.my_tempdir, "mushroom.csv")
+        obj_key = os.path.join(self.s3_prefix, "csvs", "mushroom.csv")
+        self.s3_client.download_file(self.bucket, obj_key, fname)
+        sf_from_disk = SFrame(fname)
+        s3_url = os.path.join("s3://", self.bucket, obj_key)
+        sf_from_s3 = SFrame(s3_url)
+        self._test_sframe_equal(sf_from_disk, sf_from_s3)
+
+    @pytest.mark.parametrize(
+        "folder", ["big_sframe_od", "medium_sframe_ac", "small_sframe_dc"]
+    )
+    def test_s3_sframe_download_and_upload(self, folder):
+        tmp_folder = os.path.join(self.my_tempdir, folder)
+        try:
+            shutil.rmtree(tmp_folder)
+        except FileNotFoundError:
+            pass
+        read_prefix = os.path.join(self.s3_prefix, "sframes/")
```

Comment:
I'm not really understanding the complexity in the below code. Comments would probably help.

If you want to test S3 downloads and uploads, why can't you just do something like the code below?
```python
s3_path = self.s3_prefix + '/sf'
local_sf.save(s3_path)
downloaded_sf = tc.load_sframe(s3_path)
_assert_sframe_equal(local_sf, downloaded_sf)
```

Reply:
1. I download from a read-only source to verify my download. 
2. I upload to a scratch space where I can mess up at my will. Since I verified my download, it the round trip succeeds, I'm confident to say the upload is right.

In this way, it indeed wastes 1 download. I can save one download as you suggested.

Reply:
I'm don't really care about saving a download. The reason I make this suggest was because it make the code simpler and easier to understand. I also think it's more like how our users are likely to use it.

Reply:
you mean, the user cares about downloading more? let's consider this scenario, you download the sframe and do a bunch of data cleaning and then you save it back. Do you think that's a reasonable use case or not?

Reply:
Yes, that's very reasonable.